<!Doctype html>
<html>
    <head>
        <meta charset="UTF-8" />
        <title>Seqphase</title>
        <link rel="stylesheet" href="style.css"/>
    </head>
    <body>
        <a href="https://github.com/eeg-ebe/eeg-ebe.github.io/tree/master/SeqPHASE">
            <svg width="80" height="80" viewBox="0 0 250 250" style="fill:white; color:black; z-index: 51; position: absolute; top: 0; right: 0; border: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path>
            </svg>
        </a>
        <div class="contentdiv">
            <h1>SeqPHASE</h1>
            <p id="cite">Flot (2010) <strong>SeqPHASE: a web tool for interconverting PHASE input/output files and FASTA sequence alignments </strong> <em>Molecular Ecology Ressources</em> <strong>10</strong> (1): 162-166 [<a href="http://dx.doi.org/10.1111/j.1755-0998.2009.02732.x">link</a>]</p><br>
            <div class="navigation">
                <a href="index.html">Back to startpage</a>
                <a href="step1.html">Start with step1</a>
                <a href="mailto: jflot@ulb.ac.be">E-Mail the author</a>
            </div>
            <h2>Frequently Asked Questions</h2>
            <div class="faqs">
                <div class="faq">
                    <a name="q1"></a>
                    <div class="question">Q1: What is SeqPHASE?</div>
                    <div class="answer">SeqPHASE is a web software that converts FASTA sequence alignments into the input file format requested for haplotyping using PHASE (<a href="http://stephenslab.uchicago.edu/software.html">http://stephenslab.uchicago.edu/software.html</a>) and transforms PHASE output files back into FASTA. SeqPHASE is particularly geared towards molecular ecologists who use PHASE to find out the haplotypes of nuclear sequence markers from direct sequencing, but may be useful to others as well.</div>
                </div>
                <div class="faq">
                    <a name="q2"></a>
                    <div class="question">Q2: How does it work?</div>
                    <div class="answer">Symbols accepted in the alignments are A (adenine), T (thymine), C (cytosine), G (guanine), W (adenine or cytosine), Y (cytosine or thymine), K (guanine or thymine), M (adenine or cytosine), S (cytosine or guanine), R (adenine or guanine), N and ? (missing information), and - (indel). After the user hits the Submit button on the web form, SeqPHASE starts by verifying that all sequences have the same length, that they contain only authorized symbols and that sequences all have different names. It then removes constant positions, inventories variable positions for which more than two possible nucleotides are found and creates up to three files: one .inp necessary to run PHASE, one .known detailing which phases are known (if any) and one .const recording the constant positions that were removed from the alignment (if any). Since PHASE does not accept letters for multistate characters, nucleotides are written into the .inp file as numbers based on alphabetical order as a mnemonic: -1 or ? for missing information (depending on whether the position displays two or more than two different nucleotides), 0 for indel, 1 for A, 2 for C, 3 for G and 4 for T. Depending on the properties of the dataset submitted, a suitable command syntax for running PHASE is suggested on the output webpage; for more information and other possible options, please refer to the PHASE 2.1 documentation (<a href="http://stephenslab.uchicago.edu/assets/software/phase/instruct2.1.pdf">http://stephenslab.uchicago.edu/assets/software/phase/instruct2.1.pdf</a>)</div>
                </div>
                <div class="faq">
                    <a name="q3"></a>
                    <div class="question">Q3: Why does SeqPHASE take up to three alignments to generate PHASE input files?</div>
                    <div class="answer">Instead of generating PHASE input files from a single FASTA alignment, which would require users to specify manually what phases are already known (for instance from cloning) and what individuals need to be phased, SeqPHASE can take as input up to three separate FASTA files: one for homozygous individuals and heterozygotes to be phased (with one sequence per individual), one for "fake haplotypes" of individuals to be phased (see Q9) and a third one for heterozygotes whose phases are already known (with two sequences per individual). In the alignment of phased heterozygotes, the names of the two sequences of each individual should differ only by their last character (e.g., 'indiv3a' and 'indiv3b'). Heterozygous individuals whose two haplotypes differ only by one substitution or insertion/deletion can be indifferently entered in the first field (with one sequence per individual), in the second field (with two sequences per individual) or in the third field (with three sequences per individual) since haplotyping is trivial in such case: this will not affect the phasing of these inviduals nor the phasing of other individuals in the dataset.</div>
                </div>
                <div class="faq">
                    <a name="q4"></a>
                    <div class="question">Q4: What is the purpose of Step 2?</div>
                    <div class="answer">In the PHASE output files, nucleotides are represented by numbers and constant positions are omitted: thus, using these files to find out the actual haplotype sequences can be a tedious and error-prone operation. This is the reason why a second script was written, taking as input the .const file generated during Step 1 and the .out or .out_pairs PHASE output file, and returning a FASTA alignment of haplotype sequences (if the .const file box is left empty, a FASTA alignment containing only the variable positions is generated). If a .out file is inputted, a list of phased haplotypes is returned as FASTA with 1-letter IUPAC indetermination code letters (R, W, M, Y, S or K, see above) at positions where phase certainty is inferior to a certain threshold (90% using PHASE default running options; this probability threshold can be modified by running PHASE using the -p and -q options, see PHASE documentation). If a .out_pairs file is inputted, a list of all possible haplotype pairs for each individual is returned as FASTA with their respective probability indicated between parentheses. Since FASTA alignments normally cannot accommodate two sequences bearing exactly the same name, the two haplotypes of each newly phased individual receive this individual's name with "a" or "b" appended.</div>
                </div>
                <div class="faq">
                    <a name="q5"></a>
                    <div class="question">Q5: What is the new "Reduced output" option in Step 2 about?</div>
                    <div class="answer">When the "Reduced output" box is ticked, homozygous individuals are represented by only one sequence (e.g., "indiv1") in the output FASTA file (which is better for some applications, such as building phylogenetic trees); otherwise, homozygous individuals are represented by two identical sequences (e.g., "indiv1a" and "indiv1b"), like in the output of PHASE. Ticking the "Reduced output" box also changes slightly the way in which the posterior probabilities are shown in the FASTA headers of the output file (this is for compatibility with a downstream application that is not yet published).</div>
                </div>
                <div class="faq">
                    <a name="q6"></a>
                    <div class="question">Q6: Is there any constraints on sequence label names?</div>
                    <div class="answer">Any sequence label name is acceptable (as long as it conforms to the FASTA general format), except in the alignment of phased haplotypes: for SeqPHASE to find out which sequences describe the two haplotypes of a given individual, the label names for these two sequences should differ only by their the last character (ex: individual23a, individual23b).</div>
                </div>
                <div class="faq">
                    <a name="q7"></a>
                    <div class="question">Q7: Is it possible to specify loci positions?</div>
                    <div class="answer">SeqPHASE was created with in mind the phasing of sequences obtained from direct sequencing of nuclear markers. As a result, it assumes that all nucleotides in the input alignment are actually contiguous and considers the locus position for each variable site to be its actual position in the alignment. However, it is easy to specify different loci positions by manually editing the PHASE input file produced by SeqPHASE (the positions of the variable sites are listed in the third line of the input file following the letter P).</div>
                </div>
                <div class="faq">
                    <a name="q8"></a>
                    <div class="question">Q8: How can I code the genotypes of length-variant heterozygotes?</div>
                    <div class="answer">Length-variant heterozygotes (LVHs) are individuals whose haplotypes are of different lengths due to the presence of one or several indels. As a result, chromatograms look fine until the first indel (with sometimes a few double peaks if there are SNPs), then display numerous double peaks due to the superposition of non-homologous bases from the two sequences (see <a href="http://dx.doi.org/10.1111/j.1471-8286.2006.01355.x">Flot et al. (2006)</a> for an explanatory figure and Figure 2 in <a href="http://dx.doi.org/10.1007/s12526-015-0319-7">Fontaneto et al. (2015)</a> for an example of such chromatogram pair). One may identify SNPs and the positions of the indels, then use SeqPHASE and PHASE to phase these polymorphisms; however, there are no IUPAC codes available to represent "A or indel", "C or indel", "G or indel" and "T or indel", which makes it impossible to represent the genotype of a length-variant heterozygote as a single string of IUPAC codes. To solve this problem, a new field was added to Step 1, in which data for individuals to be phased can be entered as pairs of fake haplotypes. Pairs of sequences entered in this field should be formatted like known phases (with names such as "indiv1a" and "indiv1b" for the two fake haplotypes of indiv1). However, since PHASE performs better when many known phases are available, the best strategy remains to phase all length-variant heterozygotes in the dataset prior to running PHASE, by using programs such as <a href="http://jfflot.mnhn.fr/champuru/">Champuru</a>, <a href="http://cowry.agri.huji.ac.il/cgi-bin/TraceHaplotyper.cgi">TraceHaplotyper</a> or <a href="http://dmitriev.speciesfile.org/indel.asp">Indelligent</a> that analyze the patterns of double peaks in the chromatograms obtained from direct sequencing.</div>
                </div>
                <div class="faq">
                    <a name="q9"></a>
                    <div class="question">Q9: How does the "fake haplotypes" field work?</div>
                    <div class="answer">Genotypes to be phased can be entered either in the 1st field of SeqPHASE (with a single sequence per individual, e.g. ATSGTYAKR) or in the 2nd field of SeqPHASE (with two "fake haplotype" sequences per individual, e.g. ATCGTCATG and ATGGTCAGA that is equivalent to the example above). Fake haplotypes are then superposed by SeqPHASE into one genotype to be phased, so it makes no difference to enter ATCGTCATG and ATGGTCAGA, or ATCGTCATA and ATGGTCAGG, or ATGGTCATG and ATCGTCAGA, etc.</div>
                </div>
                <div class="faq">
                    <a name="q10"></a>
                    <div class="question">Q10: How should I cite this program?</div>
                    <div class="answer">Flot (2010) SeqPHASE: a web tool for interconverting PHASE input/output files and FASTA sequence alignments. Molecular Ecology Ressources 10 (1): 162-166</div>
                </div>
            </div>
            <div class="navigation">
                <a href="index.html">Back to startpage</a>
                <a href="step1.html">Start with step1</a>
                <a href="mailto: jflot@ulb.ac.be">E-Mail the author</a>
            </div>
        </div>
    </body>
</html>
